Prevent uncontrolled exec config growth

crosbymichael opened the pull request:
Fixes #14444

This includes two commits:
1. This removes the exec config from the container after the command exits
   so that dead exec commands are not displayed in the container inspect.
   The commands are still kept on the daemon so that when you inspect the
   exec command, not the container, you are still able to get it's exit
   status.

This also changes the ProcessConfig to a pointer.
1. This adds an event loop for running a GC cleanup for exec command
   references that are on the daemon.  These cannot be cleaned up
   immediately because processes may need to get the exit status of the
   exec command but it should not grow out of bounds.  The loop is set to a
   default 5 minute interval to perform cleanup.

It should be safe to perform this cleanup because unless the clients are
remembering the exec id of the process they launched they can query for
the status and see that it has exited.  If they don't save the exec id
they will have to do an inspect on the container for all exec instances
and anything that is not live inside that container will not be returned
in the container inspect.

crosbymichael committed:
Remove exec config from container after exit

This removes the exec config from the container after the command exits
so that dead exec commands are not displayed in the container inspect.
The commands are still kept on the daemon so that when you inspect the
exec command, not the container, you are still able to get it's exit
status.

This also changes the ProcessConfig to a pointer.

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

crosbymichael committed:
Add GC loop to clean exec command refs on daemon

This adds an event loop for running a GC cleanup for exec command
references that are on the daemon.  These cannot be cleaned up
immediately because processes may need to get the exit status of the
exec command but it should not grow out of bounds.  The loop is set to a
default 5 minute interval to perform cleanup.

It should be safe to perform this cleanup because unless the clients are
remembering the exec id of the process they launched they can query for
the status and see that it has exited.  If they don't save the exec id
they will have to do an inspect on the container for all exec instances
and anything that is not live inside that container will not be returned
in the container inspect.

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

crosbymichael commented:
@bprashanth and @thockin do you think this fix will address your specific issues around CPU and memory? 

Thanks for checking it out.

thockin commented:
To be clear, the GC will nuke old exec info after O(minutes)?  For example, we run a health-check every 10 seconds by exec.  That means the daemon will hold (60/10)*5 = 30 old execs and that's all?  That seems workable, I think.

Thanks for jumping on it.

thockin commented:
Prashanth has better hands-on view than I do right now.

crosbymichael commented:
@thockin yes and the exec id list on the container inspect will be immediately removed whenever the exec process dies to cover that path as well.  Do you wont see large lists of exec ids even if they are all running.

duglin commented on daemon/exec.go:
```diff
@@ -248,3 +248,34 @@ func (d *Daemon) Exec(c *Container, execConfig *execConfig, pipes *execdriver.Pi
+			ids              = d.execCommands.List()
+		)
+		for _, id := range ids {
+			if _, exists := liveExecCommands[id]; !exists {
```
If an exec finishes just before the next "tick", won't it be deleted pretty much right away?  I'm wondering if we need to check the time the exec finished to make sure its at least as old as X seconds

duglin replied:
@crosbymichael ^^^

icecrime replied:
Mmm makes sense.

crosbymichael replied:
Because I get the container ids first before I itterate over the daemon objects this should not make the race happen because both the List calls are behind a mutex.  If the command starts and exits the lock on modifying the list will keep this from happening and it will still be viewed as alive for that tick.

duglin replied:
@crosbymichael but if something stops right before a 'tick' it will be erased immediate, right?  Isn't that too soon and racey vs the CLI which will be looking for that exec ID?

crosbymichael replied:
@duglin no, the locks will be held on the read/writes and it wont happen.  Since the baseline for the GC is the container's list and immediate deletes happen on that list it is synchronized.

duglin replied:
sorry - I'm probably being dense but I'm not following.  So, let me ask this... between each "tick" are there any locks held?  I hope not  :-)   So, if not, then any exec that dies between each gc tick will be a candidate for deletion during the next tick, right?

duglin replied:
to be clear, I'm not talking about an exec that dies between then time you get the list of containers (and their execs) and the total list of execs, I'm talking about one dying while the GC loop is waiting for the next tick.  It seems to me that those will be deleted upon the next tick - which could be too soon.

crosbymichael replied:
ic, its race proof but if clients are slow on getting an exit status this could happen.

duglin replied:
right - I think we need to only delete execs after a certain amount of time - still a bit racey but better than the risk of deleting them immediately.

duglin replied:
Another option, if there's no easy way to associate a timestamp with each one, is to do what I said earlier and keep a list of dead exec associated with each container.  The total list could go away when the container dies, and to avoid the appearance of a memory leak, we could even say that we'll only keep around the latest 10 execs in that list.  
Just a thought

duglin replied:
I would actually prefer that since the timestamp option is still racey given that we'll never know for sure how long the client will take to query the status of the exec.  With the 'dead list' approach they can take as long as they want, but it forces a model where they don't spin up a ton of execs at once - more of a "one at a time model".  Which I would think you'd prefer because it limits what exec can be used for.

duglin replied:
@crosbymichael what do you think about this patch?  https://github.com/docker/docker/compare/master...duglin:ExecClean?expand=1

crosbymichael replied:
@duglin I'm not a big fan of having yet another list with these things.  we already have one too many right now.  Maybe i'll make something local so things are queued and removed on the next tick after they are queued but idk,  this is not going to be a common path that we hit

duglin replied:
@crosbymichael the problem with time-based stuff is that its inherently racey because inevitably someone will complain about it happening too fast. This puts control in the hands of the user - we just set the boundaries - 10 execs in this case.

crosbymichael replied:
i actually think the time based approach is better and it's much simpler, we can iterate on this later if it turns out to be a bad idea.  since we are already timebased on the GC we will just not remove them if they exited less than 60 seconds before the gc loop.

icecrime commented:
LGTM

bprashanth commented:
Thanks for the quick turn around on this!
Suggest clarifying the remote api docs around execids returned on inspect:  https://docs.docker.com/reference/api/docker_remote_api/

bprashanth commented on daemon/exec.go:
```diff
@@ -248,3 +248,34 @@ func (d *Daemon) Exec(c *Container, execConfig *execConfig, pipes *execdriver.Pi
 	return exitStatus, err
 }
+
+// execCommandGC starts a go routine on a ticker to clean up the daemon references
```
nit: this doesn't start the goroutine right?

crosbymichael replied:
no

crosbymichael committed:
Use mark and sweep for exec command removal

This takes the final removal for exec commands in two steps.  The first
GC tick will mark the exec commands for removal and then the second tick
will remove the config from the daemon.

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

crosbymichael commented:
@duglin I changed it to a mark and sweep model for the final removal.  See the last commit for my changes.

duglin commented on daemon/exec.go:
```diff
@@ -248,3 +249,37 @@ func (d *Daemon) Exec(c *Container, execConfig *execConfig, pipes *execdriver.Pi
+// and running inside a container.
+func (d *Daemon) containerExecIds() map[string]struct{} {
+	ids := map[string]struct{}{}
+	for _, c := range d.containers.List() {
```
Since d.containers uses sync.Mutex instead of sync.RWMutex the entire d.containers struct/map will be locked during this loop for reads and writes.  Does that concern you at all?  I'm wondering if we should switch it to a RWMutex and then List() can just do a RLock() instead of Lock() to be a little less blocky.

crosbymichael replied:
no, it should happen quickly, we can optimize later if it's an issue

crosbymichael commented:
ping @jfrazelle @LK4D4

jfrazelle commented:
LGTM
